### PR TITLE
DMAC

### PR DESCRIPTION
**Pull request by laanwj**
Source: `src/drivers/include/dmac.h`.

### CODE REVIEW

**Comment by laanwj** [comment]
I've moved the `<size>64</size>` down into all registers, which seems to work. I am confused by the generated structure though:
```rust
    #[doc = "0x100 - Channel configuration"]
    pub channel0: CHANNEL,
    _reserved2: [u8; 248usize],
    #[doc = "0x200 - Channel configuration"]
    pub channel1: CHANNEL,
    _reserved3: [u8; 248usize],
    #[doc = "0x300 - Channel configuration"]
    pub channel2: CHANNEL,
    _reserved4: [u8; 248usize],
    #[doc = "0x400 - Channel configuration"]
    pub channel3: CHANNEL,
    _reserved5: [u8; 248usize],
    #[doc = "0x500 - Channel configuration"]
    pub channel4: CHANNEL,
    _reserved6: [u8; 248usize],
    #[doc = "0x600 - Channel configuration"]
    pub channel5: CHANNEL,
```
How can this be correct? I've padded `CHANNEL` to 0x100 bytes, as far as I know! It should be generating an array `channel[6]`.

**Comment by burrbull** [comment]
You can add _reserved to end of cluster like I did in i2s.

**Comment by laanwj** [comment]
I did, but even if that failed there would be no way there's 248 bytes left. THat'd mean the entire structure is only 8 bytes (did all the fields land on top of each other?).

File: k210.svd [hunk 1]
```diff
@@ -1014,13 +1014,1315 @@
...
+                        <!-- reserved bit 1..63 -->
+                    </fields>
                 </register>
+                <!-- reserved: 0x60..0xff -->
+                <cluster>
+                    <name>channel[%s]</name>
+                    <description>Channel configuration</description>
+                    <dim>6</dim>
+                    <addressOffset>0x100</addressOffset>
+                    <size>64</size>
```
Comment (burrbull) [inline k210.svd]: Try to delete this line.

**Comment by burrbull** [comment]
Looks like bug in svd2rust here:
https://github.com/rust-embedded/svd2rust/blob/20d8bd4c7554c5b9ced6c1c6941a81747274e9e4/src/generate/peripheral.rs#L602

File: k210.svd (hunk 1 above)
Comment (laanwj) [reply k210.svd]: That seems to have fixed it, thanks

File: k210.svd [hunk 2]
```diff
@@ -1014,13 +1014,1314 @@
...
             <groupName>DMAC</groupName>
             <baseAddress>0x50000000</baseAddress>
+            <!-- TODO <size>64</size> -->
             <registers>
-                <!-- TODO -->
                 <register>
-                    <name>dummy</name>
-                    <description>Dummy register: this peripheral is not implemented yet</description>
+                    <name>id</name>
+                    <description>ID Rgister</description>
```
Comment (Disasm) [inline k210.svd]:
```suggestion
                    <description>ID Register</description>
```

File: k210.svd [hunk 3]
```diff
@@ -1014,13 +1014,1314 @@
...
+                            <bitRange>[2:2]</bitRange>
+                        </field>
+                        <field>
+                            <name>slvif_wronhold_err</name>
+                            <description>Clear slvif_wronhold_err interrupt in com_intstatus</description>
+                            <bitRange>[3:3]</bitRange>
+                        </field>
+                        <!-- reserved bit 4..7 -->
+                        <field>
+                            <name>slvif_undefinedreg_dec</name>
```
Comment (Disasm) [inline k210.svd]:
```suggestion
                            <name>slvif_undefinedreg_dec_err</name>
```

File: k210.svd [hunk 4]
```diff
@@ -1014,13 +1014,1314 @@
...
+                            <description>Clear slvif_undefinedreg_dec in com_intstatus</description>
```
Comment (Disasm) [inline k210.svd]:
```suggestion
                            <description>Clear slvif_undefinedreg_dec_err in com_intstatus</description>
```

File: k210.svd [hunk 5]
```diff
@@ -1014,13 +1014,1314 @@
...
+                            <bitRange>[2:2]</bitRange>
+                        </field>
+                        <field>
+                            <name>slvif_wronhold_err</name>
+                            <description>Slave Interface Common Register Write On Hold Error</description>
+                            <bitRange>[3:3]</bitRange>
+                        </field>
+                        <!-- reserved bit 4..7 -->
+                        <field>
+                            <name>slvif_undefinedreg_dec</name>
```
Comment (Disasm) [inline k210.svd]:
```suggestion
                            <name>slvif_undefinedreg_dec_err</name>
```

File: k210.svd [hunk 6]
```diff
@@ -1014,13 +1014,1314 @@
...
```
Comment (Disasm) [inline k210.svd]:
```suggestion
                            <name>slvif_undefinedreg_dec_err</name>
```

File: k210.svd [hunk 7]
```diff
@@ -1014,13 +1014,1314 @@
...
```
Comment (Disasm) [inline k210.svd]:
```suggestion
                            <name>slvif_undefinedreg_dec_err</name>
```

File: k210.svd [hunk 8]
```diff
@@ -1014,13 +1014,1314 @@
...
+                                <bitRange>[4:4]</bitRange>
+                                <enumeratedValues>
+                                    <name>INCREMENT</name>
+                                    <enumeratedValue>
+                                        <name>increment</name>
+                                        <description>Increment address</description>
+                                        <value>0</value>
+                                    </enumeratedValue>
+                                    <enumeratedValue>
+                                        <name>nochange</name>
```
Comment (Disasm) [inline k210.svd:1521]:
```suggestion
                                        <name>no_change</name>
```

File: k210.svd [hunk 9]
```diff
@@ -1014,13 +1014,1314 @@
...
+                            <field>
+                                <name>arlen_en</name>
+                                <description>Source burst length enable</description>
+                                <bitRange>[38:38]</bitRange>
+                            </field>
+                            <field>
+                                <name>arlen</name>
+                                <description>Source burst length</description>
+                                <bitRange>[46:39]</bitRange>
+                                <enumeratedValues derivedFrom="BURST_LENGTH" />
```
Comment (Disasm) [inline k210.svd]: Are you sure that `arlen` has the same type as `src_msize`? I can't find anything about it and `arlen` is 8-bit field and `src_msize` is 4-bit, which is strange.
Same question for `awlen`.

File: k210.svd [hunk 10]
```diff
@@ -1014,13 +1014,1314 @@
...
+                            </field>
+                            <!-- reserved bit 59..61 -->
+                            <field>
+                                <name>shadowreg_or_lli_last</name>
+                                <description>Last shadow linked list item (indicates shadowreg/LLI content is the last one)</description>
+                                <bitRange>[62:62]</bitRange>
+                            </field>
+                            <field>
+                                <name>shadowreg_or_lli_valid</name>
+                                <description>last shadow linked list item valid (indicate shadowreg/LLI content is  valid)</description>
```
Comment (Disasm) [inline k210.svd]:
```suggestion
                                <description>last shadow linked list item valid (indicate shadowreg/LLI content is valid)</description>
```

File: k210.svd [hunk 11]
```diff
@@ -1014,13 +1014,1314 @@
...
+                                        <description>Continuous multi-block type</description>
+                                        <value>0</value>
+                                    </enumeratedValue>
+                                    <enumeratedValue>
+                                        <name>reload</name>
+                                        <description>Reload multi-block type</description>
+                                        <value>1</value>
+                                    </enumeratedValue>
+                                    <enumeratedValue>
+                                        <name>shadowregister</name>
```
Comment (Disasm) [inline k210.svd]:
```suggestion
                                        <name>shadow_register</name>
```

File: k210.svd [hunk 12]
```diff
@@ -1014,13 +1014,1314 @@
...
+                                        <description>Shadow register based multi-block type</description>
+                                        <value>2</value>
+                                    </enumeratedValue>
+                                    <enumeratedValue>
+                                        <name>linkedlist</name>
```
Comment (Disasm) [inline k210.svd]:
```suggestion
                                        <name>linked_list</name>
```

File: k210.svd [hunk 13]
```diff
@@ -1014,13 +1014,1314 @@
...
+                                    </enumeratedValue>
+                                    <enumeratedValue>
+                                        <name>block_transfer</name>
+                                        <description>Duration of channel is locked for current block transfer</description>
+                                        <value>1</value>
+                                    </enumeratedValue>
+                                    <enumeratedValue>
+                                        <name>transaction</name>
+                                        <description>Duration of channel is locked for current transaction</description>
+                                        <value>2</value>
```
Comment (Disasm) [inline k210.svd:1879]: Hmm, can't find this in recent SDK. Where did you get this?

**Review by Disasm — CHANGES_REQUESTED** [review]
Thank you!
I found a few debatable things. Could you provide any feedback on them?

File: src/lib.rs [hunk 14]
```diff
@@ -724,7 +724,6 @@ impl Peripherals {
     }
     #[doc = r" Unchecked version of `Peripherals::take`"]
     pub unsafe fn steal() -> Self {
-        debug_assert!(!DEVICE_PERIPHERALS);
```
Comment (laanwj) [inline src/lib.rs:727]: Hm this change is pretty silly here, would probably make sense to do this in a separate commit/PR (it's due to newer svd2rust commit I think?)

File: k210.svd (hunk 13 above)
Comment (laanwj) [reply k210.svd:1879]: Will check

File: src/lib.rs (hunk 14 above)
Comment (Disasm) [reply src/lib.rs:727]: Yes, it's from latest svd2rust. I don't think it makes sense to commit it as a separate PR.

File: k210.svd (hunk 13 above)
Comment (laanwj) [reply k210.svd:1879]: This is `DMAC_LOCK_LEVEL_TRANSACTION` in `dmac_lock_level_t`

File: k210.svd (hunk 9 above)
Comment (laanwj) [reply k210.svd]: True, I have no idea here, no type seems to be defined for 'burst length' only 'burst transaction length'

File: src/lib.rs (hunk 14 above)
Comment (laanwj) [reply src/lib.rs:727]: I'll leave it in here then (and make sure it doesn't leak into other PRs)

**Comment by Disasm** [comment]
Sorry, I forgot about this PR. Thanks for the corrections!
